mathics/builtin/comparison.py

```diff
@@ -376,13 +377,10 @@ def apply_other(self, args, evaluation):
 class _ComparisonOperator(_InequalityOperator):
     "Compares arguments in a chain e.g. a < b < c compares a < b and b < c."
 
-    def apply(self, items, evaluation):
-        "%(name)s[items___]"
-        items_sequence = items.get_sequence()
-        if len(items_sequence) <= 1:
-            return SymbolTrue
-        items = self.numerify_args(items, evaluation)
-        wanted = operators[self.get_name()]
+    @staticmethod
```

Comment:
This is basically taking what I wrote about `evaluation_fast()` and doing this directly in Python code. 

Sure, this is fine too, and probably we should do more of this as well. 

But as before I really want to more benchmarks such as for inequality. 

And the `benchmarks.py` routine should be moved over to the benchmarks repository. 

It has some nice features, but I note that like everything else it is a monolith that isn't going to scale as is. You run all the tests all the time as opposed to having tests isolated and then combining them in a single report if that's what you want to do.

Reply:
> This is basically taking what I wrote about `evaluation_fast()` and doing this directly in Python code.
> 
yep. The problem is that I do not know how that function could be implemented. I think that the best is to have Python functions (or maybe C functions?) for functions that are called for often, avoiding the Mathics-parser calling mechanism.



> Sure, this is fine too, and probably we should do more of this as well.
> 
> But as before I really want to more benchmarks such as for inequality.
> 
> And the `benchmarks.py` routine should be moved over to the benchmarks repository.
> 
> It has some nice features, but I note that like everything else it is a monolith that isn't going to scale as is. You run all the tests all the time as opposed to having tests isolated and then combining them in a single report if that's what you want to do.

OK with the benchmarks. When we have a more complete interface, I can try to move the benchmark tests there.

Reply:
I like the idea of implementing something in C. The Python C API is hard to use (I think) but is very complete. It can be done in a way that the overhead is almost 0 (if not 0). That can be reached because we can create Python objects from C, not just convert C object to Python.

If it be done, what would be written in C? (at least to start)

Do you and @rocky think that worths translating mathics-scanner to C?

Reply:
I think that mathics-scanner is OK as it is. What is very slow is the pattern matching over S-Expressions. This is what should be improved, maybe reimplementing it in C, or just by looking at a better way of representing Expressions.

Reply:
For example, look at https://github.com/corywalker/expreduce
This is another implementation of WL, but written in Go. The implementation is much more limited, but I think the performance of the pattern matcher is much better.

Reply:
Knowing Go, that should be instantaneous. I'll look at that, but probably the code is too different from Mathics to be useful. I don't know how to split pattern matching. What I already thought was to write the `Buitin` class in C, but that was very hard. Anyway, I'll try that again another day.

The problem of using Go is that it has a garbage collector, therefore, the overhead would be bigger.

Reply:
To say the truth, expreduce seems to be very close to Mathics in pattern matching. Dunno if rewriting that in C would be much faster, but with sure that would be faster. Now I have more ideas to what I should do.

Reply:
The idea was not to try to implement anything in Go, but just to have another open implementation that we can use to compare. 
On the other hand, I do not think that the problem is that `Builtin` classes are written in Python, but the way in which the methods of these classes are accessed during the evaluation.

Reply:
By what I understand all apply methods are stored in a big dict.